File: leetcode_java/src/main/java/LeetCodeJava/HashTable/ArrayPartitionI.java
package LeetCodeJava.HashTable;

// https://leetcode.com/problems/array-partition-i/

import java.util.Arrays;

/**
 *  561. Array Partition I  (a.k.a. Array Partition)
 *  Easy
 *
 *  Given an integer array nums of 2n integers, group these integers into n pairs
 *  (a1, b1), (a2, b2), ..., (an, bn) such that the sum of min(ai, bi) for all i
 *  is maximized. Return the maximized sum.
 *
 *  Example 1:
 *  Input: nums = [1,4,3,2]
 *  Output: 4
 *  Explanation: pairs are (1, 2) and (3, 4) -> min(1,2) + min(3,4) = 1 + 3 = 4
 *
 *  Example 2:
 *  Input: nums = [6,2,6,5,1,2]
 *  Output: 9
 *  Explanation: pairs (2,1), (2,5), (6,6) -> 1 + 2 + 6 = 9
 *
 *  Constraints:
 *  1 <= n <= 10^4, nums.length == 2 * n
 *  -10^4 <= nums[i] <= 10^4
 */
public class ArrayPartitionI {

    // V0
    // IDEA: GREEDY + SORT. pair adjacent elements after sorting,
    //       so every "wasted" larger element is as small as possible -> take nums[0], nums[2], ...
    /**
     * time = O(n log n)
     * space = O(1)   // ignoring sort's internal space
     */
    public int arrayPairSum(int[] nums) {
        Arrays.sort(nums);
        int res = 0;
        for (int i = 0; i < nums.length; i += 2) {
            res += nums[i];
        }
        return res;
    }

    // V1
    // IDEA: COUNTING SORT over the bounded value range [-10^4, 10^4]
    /**
     * time = O(n + r)   // r = value range size
     * space = O(r)
     */
    public int arrayPairSum_1(int[] nums) {
        final int OFFSET = 10000;
        int[] cnt = new int[2 * OFFSET + 1];
        for (int n : nums) {
            cnt[n + OFFSET]++;
        }

        int res = 0;
        boolean takeNext = true; // true -> next element scanned is a "min" of its pair
        for (int v = 0; v < cnt.length; v++) {
            while (cnt[v] > 0) {
                if (takeNext) {
                    res += v - OFFSET;
                }
                takeNext = !takeNext;
                cnt[v]--;
            }
        }
        return res;
    }
}

File: leetcode_python/Hash_table/array-partition-i.py
"""

561. Array Partition
Easy

Given an integer array nums of 2n integers, group these integers into n pairs (a_1, b_1), (a_2, b_2), ..., (a_n, b_n) such that the sum of min(a_i, b_i) for all i is maximized. Return the maximized sum.

Example 1:

Input: nums = [1,4,3,2]
Output: 4
Explanation: All possible pairings (ignoring the ordering of elements) are:
1. (1, 4), (2, 3) -> min(1, 4) + min(2, 3) = 1 + 2 = 3
2. (1, 3), (2, 4) -> min(1, 3) + min(2, 4) = 1 + 2 = 3
3. (1, 2), (3, 4) -> min(1, 2) + min(3, 4) = 1 + 3 = 4
So the maximum possible sum is 4.

Example 2:

Input: nums = [6,2,6,5,1,2]
Output: 9
Explanation: The optimal pairing is (2, 1), (2, 5), (6, 6). min(2, 1) + min(2, 5) + min(6, 6) = 1 + 2 + 6 = 9.

Constraints:

1 <= n <= 10^4
nums.length == 2 * n
-10^4 <= nums[i] <= 10^4

"""

# time = O(r), r is the range size of the integers
# space = O(r)
# Given an array of 2n integers, your task is to group these integers into n pairs of integer,
# say (a1, b1), (a2, b2), ..., (an, bn) which makes sum of min(ai, bi) for all i from 1 to n as large as possible.
#
# Example 1:
# Input: [1,4,3,2]
#
# Output: 4
# Explanation: n is 2, and the maximum sum of pairs is 4.
# Note:
# n is a positive integer, which is in the range of [1, 10000].
# All the integers in the array will be in the range of [-10000, 10000].


# V0 

# V1 
# http://bookshadow.com/weblog/2017/04/23/leetcode-array-partition-i/
# time = O(n log n)
# space = O(n)
class Solution(object):
    def arrayPairSum(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        return sum(sorted(nums)[::2])

# V2
# time = O(r)  # r = range size of the integers
# space = O(r)
class Solution(object):
    def arrayPairSum(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        LEFT, RIGHT = -10000, 10000
        lookup = [0] * (RIGHT-LEFT+1)
        for num in nums:
            lookup[num-LEFT] += 1
        r, result = 0, 0
        for i in xrange(LEFT, RIGHT+1):
            result += (lookup[i-LEFT] + 1 - r) / 2 * i
            r = (lookup[i-LEFT] + r) % 2
        return result

# time = O(n log n)
# space = O(1)
class Solution2(object):
    def arrayPairSum(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums.sort()
        result = 0
        for i in range(0, len(nums), 2):
            result += nums[i]
        return result

# time = O(n log n)
# space = O(n)
class Solution3(object):
    def arrayPairSum(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums = sorted(nums)
        return sum([nums[i] for i in range(0, len(nums), 2)])
